ospf6d: rework BFD integration and add BFD profiles support

rzalamena opened the pull request:
Summary
---

Rework BFD integration code in OSPFv3 and add support for BFD profiles.

Extra: reduce `bfd-profiles-topo1` topology convergence time by 10 seconds.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/be6137e1573641a1d46ccfdab4d2efe7/raw/ee62bb6fdd3ec82e8bd8aa238adf28d31ea8efd5/cr_8495_1618603407.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospf6d/ospf6_bfd.c b/ospf6d/ospf6_bfd.c
index 3b9f9fd88..7ca1326c1 100644
--- a/ospf6d/ospf6_bfd.c
+++ b/ospf6d/ospf6_bfd.c
@@ -168,8 +168,7 @@ void ospf6_bfd_write_config(struct vty *vty, struct ospf6_interface *oi)
 	}
 }
 
-DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd,
-      "ipv6 ospf6 bfd [profile BFDPROF]",
+DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd, "ipv6 ospf6 bfd [profile BFDPROF]",
       IP6_STR OSPF6_STR
       "Enables BFD support\n"
       "BFD Profile selection\n"
@@ -191,7 +190,8 @@ DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd,
 	oi->bfd_config.enabled = true;
 	if (argc > prof_idx) {
 		XFREE(MTYPE_TMP, oi->bfd_config.profile);
-		oi->bfd_config.profile = XSTRDUP(MTYPE_TMP, argv[prof_idx]->arg);
+		oi->bfd_config.profile =
+			XSTRDUP(MTYPE_TMP, argv[prof_idx]->arg);
 	}
 
 	ospf6_bfd_reg_dereg_all_nbr(oi, true);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/14f4142e565058bc17291df83b1d6c60/raw/c435b1ac1aa1512ec5bf6c31b520dbee651f0e73/cr_8495_1618603560.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospf6d/ospf6_bfd.c b/ospf6d/ospf6_bfd.c
index 32c200b50..7ca1326c1 100644
--- a/ospf6d/ospf6_bfd.c
+++ b/ospf6d/ospf6_bfd.c
@@ -168,8 +168,7 @@ void ospf6_bfd_write_config(struct vty *vty, struct ospf6_interface *oi)
 	}
 }
 
-DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd,
-      "ipv6 ospf6 bfd [profile BFDPROF]",
+DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd, "ipv6 ospf6 bfd [profile BFDPROF]",
       IP6_STR OSPF6_STR
       "Enables BFD support\n"
       "BFD Profile selection\n"

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rzalamena commented:
Rebased to fix a formatting bit pointed out by polychaeta.

idryzhov commented on ospf6d/ospf6_bfd.c:
```diff
@@ -254,70 +126,86 @@ static int ospf6_bfd_interface_dest_update(ZAPI_CALLBACK_ARGS)
 #endif /* ! HAVE_BFDD */
-		vty_out(vty, " ipv6 ospf6 bfd\n");
+	{
+		if (oi->bfd_config.profile)
```
Profile config should be displayed regardless of the other config.

rzalamena replied:
You mean the BFD should be displayed twice? One to enable BFD and a second time just for the profile?

idryzhov replied:
Yes. The config is displayed this way in BGP/OSPF, I think it should be consistent throughout the daemons.

rzalamena replied:
Fixed

idryzhov commented on ospf6d/ospf6_bfd.c:
```diff
@@ -118,134 +87,37 @@ void ospf6_bfd_trigger_event(struct ospf6_neighbor *on, int old_state,
-
-			ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_UPDATE);
-		}
+		ospf6_bfd_info_nbr_create(oi, on);
+		bfd_sess_enable(on->bfd_session, enable);
+		bfd_sess_install(on->bfd_session);
```
This is for the "enable" case. The "disable" case is not processed. We should free the BFD session in this case.

rzalamena replied:
This is fine... this is a generalization to avoid code duplication.

If `enable = false` then install will do nothing:

```c
void bfd_sess_install(struct bfd_session_params *bsp)
{
	/* Don't attempt to install/update when disabled. */
	if (!bsp->enabled)
		return;

	bsp->lastev = BSE_INSTALL;
	thread_add_event(bsglobal.tm, _bfd_sess_send, bsp, 0, &bsp->installev);
}
```

idryzhov replied:
There should be code to free the BFD session somewhere. Currently the session is never freed.

rzalamena replied:
The session is `free()`d when:

1. Session is removed ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_neighbor.c#L153 )
2. Interface is configuration is removed ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_interface.c#L246 )
3. On shutdown when interface configurations are purged ( https://github.com/opensourcerouting/frr/blob/ospfv3-bfd-rework/ospf6d/ospf6_main.c#L96 )

I think you mean when the BFD configuration is removed... In this case I need to think how to `free` the memory without just deactivating it. Or did I miss somewhere else?

idryzhov replied:
Sure, sorry, I mean when the config is removed. We do this in other daemons.
I think just adding a separate case to `ospf6_bfd_reg_dereg_all_nbr` should be enough:
```
		if (enable) {
			ospf6_bfd_info_nbr_create(oi, on);
			bfd_sess_enable(on->bfd_session, enable);
			bfd_sess_install(on->bfd_session);
		} else {
			bfd_sess_free(&on->bfd_session);
		}
```

rzalamena replied:
Fixed

idryzhov commented on ospf6d/ospf6_bfd.c:
```diff
@@ -44,72 +44,41 @@
+						&on->linklocal_addr);
+		}
+
+		bfd_sess_enable(on->bfd_session, true);
```
Why do we need to change the "enable" flag here? This is the configurational value.

rzalamena replied:
The difference from PIM to OSPFv3 is that OSPFv3 has BFD information in the peer `show` commands.

In PIM if we have the BFD session data structure or not it does not matter because it is never shown (we can free the memory), however on OSPFv3 it displays the configuration and uptime.

The difference between `bfd_sess_uninstall` and `bfd_sess_enable(..., false)` is that the former will keep the session "disabled" (will be skipped/ignored) even on the case of daemon reconnection. If the session is only "uninstalled" it might get reinstalled if `zebra` or `bfdd` restarts. `bfd_sess_uninstall` is more of a action than a state and `bfd_sess_enable(..., false)` is the same as `uninstall` plus keeping it disabled.

```c
void bfd_sess_enable(struct bfd_session_params *bsp, bool enable)
{
	/* Remove the session when disabling. */
	if (!enable)
		_bfd_sess_remove(bsp);

	bsp->enabled = enable;
}
```

And here is why just uninstalling (`bfd_sess_uninstall`) is insufficient: https://github.com/FRRouting/frr/blob/master/lib/bfd.c#L1064

In conclusion: we disable the session so it gets uninstalled (and never gets installed again) and to keep the session data structure so we can keep track of current configuration + uptime.

I hope this answers this and the other questions.

idryzhov replied:
I am actually comparing this with OSPF implementation, not PIM, and the implementation is very different.
In `ospf_bfd_trigger_event` we only do install/uninstall without touching the enable flag.
If I understand your explanation correctly then we might have a problem in ospfd.

rzalamena replied:
I tested it locally and both `ospfd` and `ospf6d` delete the neighbor data structure along with `bfd_session` when session goes down... I did the extra effort in `ospf6d` but the gain was not that much.

When the session goes down the neighbor gets deleted and the session information doesn't get shown anyway. The only difference is: meanwhile the session doesn't go to state `Full` it shows the peer as `down`:

```
 Neighbor 10.254.254.5%r4-eth1
    Area 0.0.0.0 via interface r4-eth1 (ifindex 3)
    His IfIndex: 2 Link-local address: fe80::a44c:53ff:fe5e:daaf
    State Full for a duration of 00:00:00
    His choice of DR/BDR 10.254.254.5/0.0.0.0, Priority 1
    DbDesc status: Slave SeqNum: 0x254d0400
    Summary-List: 0 LSAs
    Request-List: 0 LSAs
    Retrans-List: 1 LSAs
      [Router Id:0.0.0.0 Adv:10.254.254.4]
    0 Pending LSAs for DbDesc in Time 00:00:00 [thread off]
    0 Pending LSAs for LSReq in Time 00:00:04 [thread on]
    0 Pending LSAs for LSUpdate in Time 00:00:04 [thread on]
    0 Pending LSAs for LSAck in Time 00:00:00 [thread off]
  BFD: Type: single hop
  Detect Multiplier: 3, Min Rx interval: 300, Min Tx interval: 300
  Status: Down, Last update: 0:00:00:01
```

rzalamena replied:
Fixed: BFD session data structure is always around except when BFD is not configured. Since `ospfd` and `ospf6d` does not keep neighbor data structures around when session goes away it doesn't make sense to do the extra state effort.

idryzhov commented on ospf6d/ospf6_bfd.c:
```diff
@@ -44,72 +44,41 @@
+		bfd_sess_install(on->bfd_session);
+	} else if (old_state >= OSPF6_NEIGHBOR_TWOWAY
+		   && state < OSPF6_NEIGHBOR_TWOWAY)
+		bfd_sess_enable(on->bfd_session, false);
```
Same question. Should be `bfd_sess_uninstall`.

rzalamena replied:
Fixed

idryzhov commented on ospf6d/ospf6_bfd.c:
```diff
@@ -118,134 +87,37 @@ void ospf6_bfd_trigger_event(struct ospf6_neighbor *on, int old_state,
 		if (on->state < OSPF6_NEIGHBOR_TWOWAY) {
-			if (command == ZEBRA_BFD_DEST_DEREGISTER)
-				bfd_info_free(
-					(struct bfd_info **)&on->bfd_info);
+			bfd_sess_enable(on->bfd_session, false);
 			continue;
 		}
```
Why do we need to disable the session? We just don't need to call `bfd_session_install` if the OSPF session is not established yet.

idryzhov replied:
I think we need to check for `on->bfd_session` being non-NULL here.

rzalamena replied:
Fixed: session always get the `bfd_session` allocated even when not state `full`.

idryzhov reviewed (changes requested):
More questions added.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/d24b32d28e7c5e2701ec43f1d8f067a0/raw/777d3f0e1ec34249579a58a26e55400e3dcf5069/cr_8495_1619031769.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospf6d/ospf6_bfd.c b/ospf6d/ospf6_bfd.c
index b80d29578..716c94e05 100644
--- a/ospf6d/ospf6_bfd.c
+++ b/ospf6d/ospf6_bfd.c
@@ -174,8 +174,7 @@ void ospf6_bfd_write_config(struct vty *vty, struct ospf6_interface *oi)
 			oi->bfd_config.profile);
 }
 
-DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd,
-      "ipv6 ospf6 bfd [profile BFDPROF]",
+DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd, "ipv6 ospf6 bfd [profile BFDPROF]",
       IP6_STR OSPF6_STR
       "Enables BFD support\n"
       "BFD Profile selection\n"

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rzalamena commented:
Rebased and added a new commit to simplify BFD integration code:

Check if the configuration parameters hasn't changed to avoid reinstalling a session without real changes (plus more docs).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/13aafeea2127f963f4fa891892e7b8c1/raw/dd9d12368756c89cd886232c563877c76232c8fa/cr_8495_1619118782.diff | git apply
```

</p>

<p>

```diff
diff --git a/ospf6d/ospf6_bfd.c b/ospf6d/ospf6_bfd.c
index ba8c398de..13022a7a2 100644
--- a/ospf6d/ospf6_bfd.c
+++ b/ospf6d/ospf6_bfd.c
@@ -170,8 +170,7 @@ void ospf6_bfd_write_config(struct vty *vty, struct ospf6_interface *oi)
 			oi->bfd_config.profile);
 }
 
-DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd,
-      "ipv6 ospf6 bfd [profile BFDPROF]",
+DEFUN(ipv6_ospf6_bfd, ipv6_ospf6_bfd_cmd, "ipv6 ospf6 bfd [profile BFDPROF]",
       IP6_STR OSPF6_STR
       "Enables BFD support\n"
       "BFD Profile selection\n"

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

rzalamena commented:
CI:rerun

idryzhov commented:
BGP test failures are not related to this PR. Merging.
